kvwrapper_etcd_v3/kvwrapper_etcd_v3.go

```diff
@@ -0,0 +1,178 @@
+package kvwrapper_etcd_v3
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/behance/go-common/kvwrapper"
+	log "github.com/behance/go-logging/log"
+	etcdv3 "github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/etcdserver/api/v3rpc/rpctypes"
+)
+
+// EtcdWrapper wraps the go-etcd client so it can implement the KVWrapper interface
+type EtcdV3Wrapper struct {
+	kapi etcdv3.KV
+	cli  etcdv3.Client
+}
+
+// NewKVWrapper returns a new kvwrapper_etcd as a KVWrapper
+func (e EtcdV3Wrapper) NewKVWrapper(servers []string, username, password string) kvwrapper.KVWrapper {
+	config := etcdv3.Config{
+		Endpoints:   servers,
+		Username:    username,
+		Password:    password,
+		DialTimeout: 5 * time.Second,
+	}
+	client, err := etcdv3.New(config)
+	if err != nil {
+		panic(err)
+	}
+
+	return EtcdV3Wrapper{kapi: etcdv3.NewKV(client), cli: *client}
+}
+
+// Set sets the key = val with a ttl of ttl. If key is a path, it will be created.
+// The ttl will be handled via a lease
+// Leases are given in increments of seconds, and leases are granted in increments of seconds,
+// hence there is no need to convert to time.Duration (which is in nanoseconds) as is done in version 2
+// note! The API assumes that a ttl of 0, implies a wish that the key/value pair not expire
+// implementing that behavior in v3, by not acquiring a lease
+func (e EtcdV3Wrapper) Set(key string, val string, ttl uint64) error {
+	// acquire a lease to which we'll attach the new key/value pair
+	// lease.grant takes an int64 as the ttl, hence casting is necessary
+	log.Debug("called Set with key/value: ", key, val, " and ttl ", ttl)
+	if ttl == 0 {
+		_, put_err := e.kapi.Put(context.Background(), key, val)
+		if put_err != nil {
+			log.Warn("Could not set key in etcd.", "key", key, "err", put_err)
+			return put_err
+		}
+	} else {
+		lease, lease_err := e.cli.Grant(context.Background(), int64(ttl))
+		if lease_err != nil {
+			log.Fatal(lease_err)
+			return lease_err
+		}
+
+		log.Debug("go-common: Set - got lease", lease.ID, " with a ttl of ", lease.TTL)
+		// Insert key with a lease of ttl second TTL
+		_, put_err := e.kapi.Put(context.Background(), key, val, etcdv3.WithLease(lease.ID))
+		if put_err != nil {
+			log.Warn("Could not set key in etcd.", "key", key, "err", put_err)
+			// if the Put op failed, clean up the lease
+			_, revoke_err := e.cli.Revoke(context.Background(), lease.ID)
+			if revoke_err != nil {
+				log.Warn("Attempt to revoke lease ", lease.ID, " failed with error ", revoke_err)
```

Comment:
Minor, but this line will result in some "interesting" logging. The way log15 works, the first param will be the main log message, and from there, each two parameters will be interpreted as key, values.

So this would lead to:
```
$(lease.ID)= failed with error 
revoke_err=nil
```
In the logs. I see this is done in the kv format in almost all places in this file, but here as in a couple more places it is trying to do string concatenation.